component/web/security/src/main/java/org/exoplatform/web/security/hash/Argon2PasswordEncoder.java

```diff
@@ -0,0 +1,104 @@
+
+  private static final String     PASSWORD_SALT_USER_ATTRIBUTE = "passwordSalt128";
+
+  private static final int        ARGON2_HASHING_ITERATIONS    = 2;
+
+  private static final int        ARGON2_MEMORY_LIMIT          = 32768;
+
+  private static final int        ARGON2_PARALLEL_EXECUTIONS   = 2;
+
+  private static final int        ARGON2_HASH_LENGTH           = 32;
+
+  private static CodecInitializer codecInitializer;
+
+  @Override
+  public String encodeCredential(String username, String rawPassword) {
+    Argon2Parameters.Builder builder =
+                                     new Argon2Parameters.Builder(Argon2Parameters.ARGON2_id).withIterations(ARGON2_HASHING_ITERATIONS)
+                                                                                             .withMemoryAsKB(ARGON2_MEMORY_LIMIT)
+                                                                                             .withParallelism(ARGON2_PARALLEL_EXECUTIONS)
+                                                                                             .withSalt(getStoredSalt(username));
+    Argon2BytesGenerator generator = new Argon2BytesGenerator();
+    generator.init(builder.build());
+    byte[] hash = new byte[ARGON2_HASH_LENGTH];
+    generator.generateBytes(rawPassword.getBytes(StandardCharsets.UTF_8), hash, 0, hash.length);
+    try {
+      return getCodec().encode(Hex.encodeHexString(hash));
+    } catch (TokenServiceInitializationException e) {
+      LOG.error("Error while applying symmetrical encryption on password hash", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private byte[] getStoredSalt(String username) {
+    try {
+      AttributesManager attributesManager = getIdentitySession().getAttributesManager();
+      Attribute salt = attributesManager.getAttribute(username, PASSWORD_SALT_USER_ATTRIBUTE);
+      if (salt == null) {
+        byte[] generatedSalt = generateRandomSalt();
+        String saltString = Hex.encodeHexString(generatedSalt);
+        attributesManager.addAttribute(username, PASSWORD_SALT_USER_ATTRIBUTE, saltString);
+        return generatedSalt;
+      } else {
+        return Hex.decodeHex(((String) salt.getValue()).toCharArray());
+      }
+    } catch (Exception e) {
+      LOG.error("Error while getting stored password hash salt", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private byte[] generateRandomSalt() throws NoSuchAlgorithmException {
+    SecureRandom secureRandom = getSecureRandomInstance();
+    byte[] salt = new byte[16];
+    secureRandom.nextBytes(salt);
+    return salt;
+  }
+
+  private static AbstractCodec getCodec() throws TokenServiceInitializationException {
+    if (codecInitializer != null) {
+      return codecInitializer.getCodec();
+    }
+    PortalContainer container = PortalContainer.getInstance();
+    codecInitializer = container.getComponentInstanceOfType(CodecInitializer.class);
+    return codecInitializer.getCodec();
```

Comment:
Could we make it more readable ?
```suggestion
    if (codecInitializer == null) {
      PortalContainer container = PortalContainer.getInstance();
      this.codecInitializer = container.getComponentInstanceOfType(CodecInitializer.class);
    }
   
    return codecInitializer.getCodec();
```